python/uv/private/lock.bzl

```diff
@@ -15,22 +15,258 @@
+{attr}`pip.parse.experimental_index_url`.
+""",
+        ),
+        "output": attr.output(
+            mandatory = False,
+            doc = """\
+The output file to create that can then be synced to the source tree.
+""",
+        ),
+        "python_version": attr.string(
+            doc = """\
+FIXME @aignas 2025-03-13: this needs to be a transition field. How do I do it?
+""",
+        ),
+        "srcs": attr.label_list(
+            mandatory = True,
+            allow_files = True,
+            doc = """\
+The sources that will be used. Add all of the files that would be passed as
+srcs to the `uv pip compile` command.
+""",
+        ),
+        "strip_extras": attr.bool(
+            default = False,
+            doc = """\
+Currently `rules_python` requires `--no-strip-extras` to properly function, but
+sometimes one may want to not have the extras if you are compiling the
+requirements file for using it as a constraints file.
+""",
+        ),
+        "update_target": attr.string(
+            mandatory = True,
+            doc = """\
+The string to input for the 'uv pip compile'.
+""",
+        ),
+        "_template": attr.label(
+            default = "//python/uv/private:pip_compile_template",
+            cfg = "exec",
+            executable = True,
+            doc = """\
+The template to be used for 'uv pip compile'. This is either .ps1 or bash
+script depending on what the target platform is executed on.
+""",
+        ),
+    },
+    toolchains = [
+        _uv_toolchain,
+        _py_toolchain,
+    ],
+)
+
+def _lock_run_impl(ctx):
+    info = ctx.attr.lock[_LockInfo]
+    uv = info.uv
+    srcs = info.srcs
+    py_runtime = info.py
+
+    args = [
+        uv.short_path,
+        "pip",
+        "compile",
+    ] + info.args + [
+        src.short_path
+        for src in srcs
+    ] + [
+        "--python",
+        py_runtime.interpreter.short_path,
+    ]
+
+    executable = ctx.actions.declare_file(ctx.label.name)
+    ctx.actions.expand_template(
+        template = info.template,
+        substitutions = {
+            "replace": "{}/{}".format(ctx.label.package, ctx.attr.output),
+            "uv pip compile": " ".join(args),
+        },
+        output = executable,
+        is_executable = True,
+    )
+
+    runfiles = ctx.runfiles(
+        transitive_files = depset(
+            srcs + [uv],
+            transitive = [
+                py_runtime.files,
+            ],
+        ),
+    )
+    return [
+        DefaultInfo(
+            executable = executable,
+            runfiles = runfiles,
+        ),
+    ]
+
+_lock_run = rule(
+    implementation = _lock_run_impl,
+    doc = """\
+""",
+    attrs = {
+        "lock": attr.label(
+            doc = "The lock target that is doing locking in a build action.",
+            providers = [_LockInfo],
+        ),
+        "output": attr.string(
+            doc = "The output that we would be updated.",
```

Comment:
This is a path relative to something, right? Please indicate what it's relative to.